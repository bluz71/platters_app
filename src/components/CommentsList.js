import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import axios from 'axios';
import linkify from 'linkify-lite';
import simpleFormat from '../helpers/simpleFormat';
import FontAwesome from 'react-fontawesome';
import '../styles/CommentsList.css';
import { appAuth } from '../lib/appAuth';
import { API_HOST } from '../config';

const postedIn = (comment) => {
  if (comment.for === 'Album') {
    return ` posted in ${comment.artist} / `;
  }
  else {
    return ' posted in ';
  }
};

const handleDelete = (comment, onDeleteComment) => {
  if (!window.confirm('Are you sure you want to remove this comment?')) {
    return;
  }

  const deleteCommentEndPoint = `${API_HOST}${comment.delete_path}.json`;
  axios.delete(deleteCommentEndPoint, appAuth.headers())
    .then(response => {
      onDeleteComment(comment.id);
    })
    .catch(error => {
      // XXX implement
      console.log('Error');
    });
};

const renderDeleteComment = (comment, onDeleteComment) => {
  if (
    appAuth.isLoggedIn() &&
    (appAuth.currentUser().name === comment.user_name ||
      appAuth.currentUser().admin)) {
    return (
      <span
        onClick={() => handleDelete(comment, onDeleteComment)}
        className="icon"
      >
        <FontAwesome name="trash" />
      </span>
    );
  }
};

const renderComments = (comments, onDeleteComment, shortHeader) => (
  comments.map(comment =>
    <div key={comment.id} className="Comment">
      <Link to={`/comments/${comment.user_slug}`}>
        <img className="img-responsive center-block" src={comment.gravatar_url} alt={comment.user_name} />
      </Link>
      <h2>
        <Link to={`/comments/${comment.user_slug}`}>
          {comment.user_name}
        </Link>
        {!shortHeader &&
            <small>
              {postedIn(comment)}
              <Link to={{
                pathname: `/${comment.path}`,
                state: { scrollToComments: true }
              }}>{comment.name}</Link>
            </small>}
      </h2>
      {renderDeleteComment(comment, onDeleteComment)}
      <h3 dangerouslySetInnerHTML={{ __html: comment.created_at }} />
      <div dangerouslySetInnerHTML={{ __html: linkify(simpleFormat(comment.body)) }} />
    </div>
  )
);

const CommentsList = ({ comments, onDeleteComment, shortHeader }) => (
  <div className="CommentsList">
    {renderComments([...comments.values()], onDeleteComment, shortHeader)}
  </div>
);

CommentsList.defaultTypes = {
  shortHeader: false
};

CommentsList.propTypes = {
  comments:    PropTypes.instanceOf(Map),
  shortHeader: PropTypes.bool
};

export default CommentsList;
