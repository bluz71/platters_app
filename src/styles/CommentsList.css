.Comment {
  background-color: #f8f8f8;
  border: 1px solid #eee;
  border-radius: 4px;
  padding: 1em;
  margin-bottom: 1em;
}

.Comment img {
  height: 45px;
  width: 45px;
  float: left;
  border: 1px solid #eee;
  border-radius: 50%;
  margin: 0 0.75em 0 0;
}

.Comment h2 {
  display: inline;
  margin-top: 0.5em;
  font-size: 1.4em;
  color: #3c9a5f;
}

.Comment h2 a {
  text-decoration: none;
  color: #3c9a5f;
}

.Comment h2 small {
  font-size: 0.75em;
}

.Comment h3 {
  margin-top: 0.5em;
  font-size: 0.75em;
  color: #515151;
}

.Comment p {
  white-space: pre-wrap;
  word-wrap: break-word;
}

.Comment .icon {
  float: right;
  font-size: 0.95em;
  color: #515151;
  cursor: pointer;
}

.Comment .icon:hover {
  color: #3c9a5f;
}
