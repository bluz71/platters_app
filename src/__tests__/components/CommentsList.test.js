import React from 'react';
import { shallow } from 'enzyme';
import CommentsList from '../../components/CommentsList';

const comments = () => {
  return [
    {
      'id': 1,
      'body': 'First comment',
      'created_at': '<time datetime=\"2018-05-09T04:05:00Z\" data-local=\"time-ago\">May  9, 2018  4:05am</time>',
      'user_name': 'joe',
      'user_slug': 'joe',
      'gravatar_url': 'https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon',
      'for': 'Album',
      'artist': 'ABC',
      'name': 'ABC',
      'path': 'artist/abc/album/abc'
    },
    {
      'id': 1,
      'body': 'Second comment',
      'created_at': '<time datetime=\"2018-05-05T05:34:56Z\" data-local=\"time-ago\">May  5, 2018  5:34am</time>',
      'user_name': 'joe',
      'user_slug': 'joe',
      'gravatar_url': 'https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon',
      'for': 'Album',
      'artist': 'DEF',
      'name': 'DEF',
      'path': 'artist/def/album/def'
    }
  ];
};

describe('<CommentsList />', () => {
  it('renders without crashing', () => {
    shallow(<CommentsList comments={[]} />);
  });

  it('renders a list of comments', () => {
    const wrapper = shallow(<CommentsList comments={comments()} />);
    expect(wrapper).toMatchSnapshot();
  });
});
