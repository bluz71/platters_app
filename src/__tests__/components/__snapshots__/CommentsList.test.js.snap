// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentsList /> renders a list of comments 1`] = `
<div
  className="CommentsList"
>
  <VelocityTransitionGroup
    enter={null}
    enterHideStyle={
      Object {
        "display": "none",
      }
    }
    enterShowStyle={
      Object {
        "display": "",
      }
    }
    leave={
      Object {
        "animation": "fadeOut",
        "duration": 500,
      }
    }
    runOnMount={false}
  >
    <div
      className="Comment"
      key="1"
    >
      <Link
        replace={false}
        to="/comments/fred"
      >
        <img
          alt="fred"
          className="img-responsive center-block"
          src="https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon"
        />
      </Link>
      <h2>
        <Link
          replace={false}
          to="/comments/fred"
        >
          fred
        </Link>
        <small>
           posted in ABC / 
          <Link
            replace={false}
            to={
              Object {
                "pathname": "/artist/abc/album/abc",
                "state": Object {
                  "scrollToComments": true,
                },
              }
            }
          >
            ABC
          </Link>
        </small>
      </h2>
      <h3
        dangerouslySetInnerHTML={
          Object {
            "__html": "<time datetime=\\"2018-05-09T04:05:00Z\\" data-local=\\"time-ago\\">May  9, 2018  4:05am</time>",
          }
        }
      />
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>First comment</p>",
          }
        }
      />
    </div>
    <div
      className="Comment"
      key="2"
    >
      <Link
        replace={false}
        to="/comments/joe"
      >
        <img
          alt="joe"
          className="img-responsive center-block"
          src="https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon"
        />
      </Link>
      <h2>
        <Link
          replace={false}
          to="/comments/joe"
        >
          joe
        </Link>
        <small>
           posted in ABC / 
          <Link
            replace={false}
            to={
              Object {
                "pathname": "/artist/abc/album/abc",
                "state": Object {
                  "scrollToComments": true,
                },
              }
            }
          >
            ABC
          </Link>
        </small>
      </h2>
      <h3
        dangerouslySetInnerHTML={
          Object {
            "__html": "<time datetime=\\"2018-05-05T05:34:56Z\\" data-local=\\"time-ago\\">May  5, 2018  5:34am</time>",
          }
        }
      />
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>Second comment</p>",
          }
        }
      />
    </div>
  </VelocityTransitionGroup>
</div>
`;

exports[`<CommentsList /> renders a list of comments with shortHeader 1`] = `
<div
  className="CommentsList"
>
  <VelocityTransitionGroup
    enter={null}
    enterHideStyle={
      Object {
        "display": "none",
      }
    }
    enterShowStyle={
      Object {
        "display": "",
      }
    }
    leave={
      Object {
        "animation": "fadeOut",
        "duration": 500,
      }
    }
    runOnMount={false}
  >
    <div
      className="Comment"
      key="1"
    >
      <Link
        replace={false}
        to="/comments/fred"
      >
        <img
          alt="fred"
          className="img-responsive center-block"
          src="https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon"
        />
      </Link>
      <h2>
        <Link
          replace={false}
          to="/comments/fred"
        >
          fred
        </Link>
      </h2>
      <h3
        dangerouslySetInnerHTML={
          Object {
            "__html": "<time datetime=\\"2018-05-09T04:05:00Z\\" data-local=\\"time-ago\\">May  9, 2018  4:05am</time>",
          }
        }
      />
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>First comment</p>",
          }
        }
      />
    </div>
    <div
      className="Comment"
      key="2"
    >
      <Link
        replace={false}
        to="/comments/joe"
      >
        <img
          alt="joe"
          className="img-responsive center-block"
          src="https://gravatar.com/avatar/88884e5cb90c02031f031bc4760fcf69?s=80&r=pg&d=identicon"
        />
      </Link>
      <h2>
        <Link
          replace={false}
          to="/comments/joe"
        >
          joe
        </Link>
      </h2>
      <h3
        dangerouslySetInnerHTML={
          Object {
            "__html": "<time datetime=\\"2018-05-05T05:34:56Z\\" data-local=\\"time-ago\\">May  5, 2018  5:34am</time>",
          }
        }
      />
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>Second comment</p>",
          }
        }
      />
    </div>
  </VelocityTransitionGroup>
</div>
`;
